### ISSUE TITLE
RX Packets getting dropped between network layer and benchmark kernel

### ISSUE THREAD

**Issue by 108anup** [Xilinx/xup_vitis_network_example#72]
## TLDR
I wanted to know what might cause packets to get dropped between the network layer and the benchmark kernel on the RX path. (The network layer probes show rx packets that in the udp and app registers, but the benchmark kernel probes don't show any packet counts).

## Setup
I am using the benchmark kernel (VNx) on a U280 FPGA to measure throughput across a custom NIC, i.e., the kernel sends packets on port 0, these are received by custom NIC's port 1, the custom NIC echoes these packets on its port 0 (after some header modifications), these packets are routed to port 1 of VNx.
```
----                  ----------
 TX  port 0 -> port 1     RX
 VNx                  Custom NIC
 RX  port 1 <- port 0     TX
----                  ----------
```
I use traffic_generator_0_3 as PRODUCER and traffic_generator_1_1 as CONSUMER (both are on the same single U280 FPGA but different ports). network_layer_0 has the dst ip, dst mac in the arp table and src/dst ports in the socket table for sending packets to the custom NIC. The custom NIC then changes src/dst ip, src/dst mac, src/dst port and ip checksum (I don't change udp checksum as I think it is zero by default in VNx). The modified packets arrive on port 1 of VNx and are received by network_layer_1 (verified by packet counter probes, so I am assuming the packet modifications are correct?). The receiving network_layer_1 has arp table entries and socket table entries for the modified packets. **However no packets are seen on the probes for the traffic_generator_1_1.**

My code is an adapted version of the throughput benchmarking notebook (I make no modifications to VNx itself). The custom NIC is a modified version of open nic on another U280 FPGA.

### Network layer 0 (TX)
```
{'tx_path': {'arp': {'packets': 250, 'bytes': 15000, 'cycles': 2041},
  'icmp': {'packets': 0, 'bytes': 0, 'cycles': 0},
  'ethernet_header_inserter': {'packets': 1000000000,
   'bytes': 106000000000,
   'cycles': 3016116903},
  'ethernet': {'packets': 1000000250,
   'bytes': 106000015000,
   'cycles': 3989787921},
  'app': {'packets': 1000000000, 'bytes': 64000000000, 'cycles': 3016116118},
  'udp': {'packets': 1000000000, 'bytes': 92000000000, 'cycles': 3016116887}},
 'rx_path': {'ethernet': {'packets': 251,
   'bytes': 15282,
   'cycles': 5912609319},
  'packet_handler': {'packets': 250, 'bytes': 15000, 'cycles': 2041},
  'arp': {'packets': 250, 'bytes': 15000, 'cycles': 2041},
  'icmp': {'packets': 0, 'bytes': 0, 'cycles': 0},
  'udp': {'packets': 0, 'bytes': 0, 'cycles': 0},
  'app': {'packets': 0, 'bytes': 0, 'cycles': 0}}}
```

### Network layer 1 (RX)
Note, the custom NIC is slower so some packets are dropped by the custom NIC.
```
{'tx_path': {'arp': {'packets': 250, 'bytes': 15000, 'cycles': 2041},
  'icmp': {'packets': 0, 'bytes': 0, 'cycles': 0},
  'ethernet_header_inserter': {'packets': 0, 'bytes': 0, 'cycles': 0},
  'ethernet': {'packets': 250, 'bytes': 15000, 'cycles': 2041},
  'app': {'packets': 0, 'bytes': 0, 'cycles': 0},
  'udp': {'packets': 0, 'bytes': 0, 'cycles': 0}},
 'rx_path': {'ethernet': {'packets': 650019850,
   'bytes': 68902104276,
   'cycles': 5128830024},
  'packet_handler': {'packets': 650019849,
   'bytes': 59801826108,
   'cycles': 3016117975},
  'arp': {'packets': 0, 'bytes': 0, 'cycles': 0},
  'icmp': {'packets': 0, 'bytes': 0, 'cycles': 0},
  'udp': {'packets': 650019849, 'bytes': 59801826108, 'cycles': 3016117975},
  'app': {'packets': 650019849, 'bytes': 41601270336, 'cycles': 3016117974}}}
```

### Producer kernel (traffic_generator_0_3)
```
RegisterMap {
  CTRL = Register(AP_START=0, AP_DONE=0, AP_IDLE=1, AP_READY=0, AUTO_RESTART=0),
  mode = Register(value=0),
  dest_id = Register(value=12),
  number_packets = Register(value=1000000000),
  number_beats = Register(value=1),
  time_between_packets = Register(value=0),
  reset_fsm = Register(value=0),
  debug_fsms = Register(value=0),
  out_traffic_cycles = Register(value=3016116080),
  out_traffic_bytes = Register(value=64000000000),
  out_traffic_packets = Register(value=1000000000),
  in_traffic_cycles = Register(value=0),
  in_traffic_bytes = Register(value=0),
  in_traffic_packets = Register(value=0),
  summary_cycles = Register(value=0),
  summary_bytes = Register(value=0),
  summary_packets = Register(value=0),
  debug_reset = Register(value=0)
}
```

### Consumer kernel (traffic_generator_1_1)
```
RegisterMap {
  CTRL = Register(AP_START=0, AP_DONE=0, AP_IDLE=1, AP_READY=0, AUTO_RESTART=0),
  mode = Register(value=3),
  dest_id = Register(value=0),
  number_packets = Register(value=0),
  number_beats = Register(value=0),
  time_between_packets = Register(value=0),
  reset_fsm = Register(value=0),
  debug_fsms = Register(value=0),
  out_traffic_cycles = Register(value=0),
  out_traffic_bytes = Register(value=0),
  out_traffic_packets = Register(value=0),
  in_traffic_cycles = Register(value=0),
  in_traffic_bytes = Register(value=0),
  in_traffic_packets = Register(value=0),
  summary_cycles = Register(value=0),
  summary_bytes = Register(value=0),
  summary_packets = Register(value=0),
  debug_reset = Register(value=0)
}
```

**Comment by 108anup** [comment]
I think the src/dst mac and dst ip in modified packets as well as arp table entry is correct since network_layer_1 shows udp packets.

Further I think the src ip, src/dst ports in modified packets as well as socket table entry is correct since network_layer_1 shows app packets. (If I set wrong port in socket table entry then it only shows udp packets not app packets).

**Comment by 108anup** [comment]
I am assuming this line in the network_layer_bd measures the RX app packets:
`connect_bd_intf_net -intf_net bandwidth_app_in_OUT_DBG [get_bd_intf_ports M_AXIS_nl2sk] [get_bd_intf_pins bandwidth_app_in/OUT_DBG]`, i.e., by tapping `M_AXIS_nl2sk`. M_AXIS_nl2sk is connected to the switch that distributes packets to the benchmarking kernels. I think this switch just broadcasts rx packets to all traffic_generators, so ideally they should recv packets given that M_AXIS_nl2sk has activity according to the probes.

**Comment by mariodruiz** [comment]
> I wanted to know what might cause packets to get dropped between the network layer and the benchmark kernel on the RX path. (The network layer probes show rx packets that in the udp and app registers, but the benchmark kernel probes don't show any packet counts).

Short answer, bad configuration of the socket table.

How are you configuring the socket in the receiving path?

traffic_generator_1_1 will only get segments if you configure the socket with index 1. The tdest in the output stream is the index of the socket that matches. The switch uses this index to target the different  traffic_generators.

**Comment by 108anup** [comment]
Thanks your first answer fixed it. On the rx network_layer_1, I had put the entry on socket table index 13. I moved this entry to index 1 and now I can see packets on the app probes.

Just to confirm:
The tdest in axis is set as the socket table entry index. Where exactly is this in the code for my reference (I see id being set in TableHandlerResponse but where is this propagated to axis?)
And the switch in the benchmark kernel uses tdest based routing instead of broadcasting. Again for my reference, could you please point to the part of code describing this.

**Comment by mariodruiz** [comment]
dest is assigned here https://github.com/hpcn-uam/100G-fpga-network-stack-core/blob/872d53d706c04dad50fc50faeb1bb96256060ffe/hls/UDP/udp.cpp#L218

The Switch does not support broadcasting, only 1:1 routing. https://www.xilinx.com/content/dam/xilinx/support/documents/ip_documentation/axis_infrastructure_ip_suite/v1_1/pg085-axi4stream-infrastructure.pdf page 16

**Comment by mariodruiz** [comment]
@108anup,

I checked the documentation and there's an explanation on how tdest works. https://github.com/Xilinx/xup_vitis_network_example/tree/master/NetLayers#udp

If you think this can be improved please feel free to PR.
